mediawiki-xml/src/main/java/org/xwiki/contrib/mediawiki/xml/internal/input/MediaWikiInputFilterStream.java

```diff
@@ -193,14 +193,17 @@ EntityReference toEntityReference(String reference, boolean link)
             // MediaWiki actually assume the link reference is a partial URL (it just concatenate it to the base URL) so
             // we have to decode it
             try {
+                // but not the plus sign, as we are in a path section
+                pageName = pageName.replaceAll("\\+","%2B");
                 pageName = URLDecoder.decode(pageName, "UTF-8");
             } catch (UnsupportedEncodingException e) {
                 // Should never happen
             }
         }
 
-        // MediaWiki replace the white spaces with an underscore in the URL
+        // MediaWiki replaces white spaces with a single underscore in the URL
         pageName = pageName.replace(' ', '_');
+        pageName = pageName.replaceAll("_{2,}","_");
```

Comment:
You sure all underscores are replace by a single one ? The comment suggest it's only white spaces.

Reply:
In fact at least my MediaWiki instance behaves that way that it collapses several underscores into a single one.
For example in the MediaWiki export dump at:

   https://github.com/xwiki-contrib/mediawiki/blob/MEDIAWIKI-99/mediawiki-xml/src/test/resources/mediawikixml/plus-in-pagename/data/current.xml#L49

the ``[[  Page  __  __  Page  ]]`` renders as a correct link to ``Page_Page`` inside MediaWiki

Reply:
Oh, and the comment applies only to the line above. I maybe should have added a second comment for the next line.